### PR TITLE
arch:arm:core:aarch32: Basic ARMv7-A / Cortex-A support

### PR DESCRIPTION
**Pull request by ibirnbaum**
This package contains work-in-progress versions of all the components required to get initial support of ARMv7-A / 32-bit Cortex-A CPUs up and running. For now, it's intended as a basis for discussion, in particular regarding the MMU implementation and the resulting changes to the linker command files, as well as the proper integration of each of the new or modified components (e.g. proper naming / use / location of Kconfig items).

The components included are:
- Basic context switching and interrupt handling. This is basically completely parallel to the Cortex-R implementation, i.e. every #ifdef CONFIG_ARMV7_R has been OR'ed with CONFIG_ARMV7_A, and every CONFIG_CPU_CORTEX_R has been OR'ed with CONFIG_CPU_AARCH32_CORTEX_A. Interrupt handling uses the existing GICv1 driver. System timing is provided by the ARM architected timer.
- However, the timer driver required the addition of a workaround for an erratum which affects the Cortex-A9 CPU. A description of the erratum and a reference to the corresponding documentation by ARM is provided in the respective commit.
- Basic support of the ARMv7 MMU. Supports static memory area configuration at boot time as well as dynamic configuration based on run-time calls to arch_mem_map. No userspace support yet, no stack guard pages yet, all other minor caveats are described in the respective commit.
- SoC/DT and other related data for the Xilinx Zynq-7000 SoC series. This SoC family should be a suitable candidate for a first aarch32 Cortex-A target, as drivers for all relevant peripherals are available (UART, Ethernet, with slight modifications of the existing ZynqMP implementations).
- Support for the Cortex-A's On-Chip Memory (OCM) so that the location of this block of memory can be configured and data can be explicitly placed in the corresponding section. This memory can be used, for example, for uncached device DMA.
- The "qemu_cortex_a9" board definition, which simulates the Xilinx Zynq-7000. The QEMU simulation includes all features relevant for testing & QA. NOTE: at least the QEMU version that currently comes with Ubuntu 20.04, qemu-system-arm version 4.2.1, crashes with a Floating Point Exception when starting the simulation, although the FPU isn't used within the target. I've manually installed a different version of QEMU that worked before, I'll try to provide more info which QEMU versions work properly. I believe this problem to be a QEMU bug.
- A fix for interrupt-driven operation of the Xilinx PS UART driver.
- An update to the GEM Ethernet driver so that it supports the Zynq-7000 SoCs in addition to the ZynqMP SoCs.

What is NOT included (that hasn't been mentioned already above):
- FPU support. I've removed my old FPU support implementation as recent discussions hinted at an implementation for the Cortex-R VFPv3-D16 FPU being in the works, this should be expandable to the Cortex-A's FVPv3-D32 FPU.

### CODE REVIEW

**Comment by carlocaione** [comment]
On a first look LGTM (great work BTW). Can you fix the buildkite errors though?

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+	if (pt_attrs & MATTR_CACHE_INNER_WB_WA) {
+		reg_val |= ARM_MMU_TTBR_IRGN0_BIT_MP_EXT_ONLY;
+	} else if (pt_attrs & MATTR_CACHE_INNER_WT_nWA) {
+		reg_val |= ARM_MMU_TTBR_IRGN1_BIT_MP_EXT_ONLY;
+	} else if (pt_attrs & MATTR_CACHE_INNER_WB_nWA) {
+		reg_val |= ARM_MMU_TTBR_IRGN0_BIT_MP_EXT_ONLY;
+		reg_val |= ARM_MMU_TTBR_IRGN1_BIT_MP_EXT_ONLY;
+	}
+
+	__asm__ __volatile__("mcr p15, 0, %0, c2, c0, 0" : : "r"(reg_val));
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:601]: I think you can use __get_TTBR0() from the cmsis

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+
+	/* Write DACR -> all domains to client = 01b. */
+	reg_val = ARM_MMU_DACR_ALL_DOMAINS_CLIENT;
+	__asm__ __volatile__("mcr p15, 0, %0, c3, c0, 0" : : "r"(reg_val));
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:605]: __set_DACR() from cmsis?

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+
+	invalidate_tlb_all();
+
+	/* Enable the MMU and Cache in SCTLR */
+	__asm__ __volatile__("mrc p15, 0, %0, c1, c0, 0" : "=r"(reg_val));
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:610]: __get_SCTLR()

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 4]
```diff
@@ -0,0 +1,800 @@
...
+	reg_val |= ARM_MMU_SCTLR_AFE_BIT;
+	reg_val |= ARM_MMU_SCTLR_ICACHE_ENABLE_BIT;
+	reg_val |= ARM_MMU_SCTLR_DCACHE_ENABLE_BIT;
+	reg_val |= ARM_MMU_SCTLR_MMU_ENABLE_BIT;
+	__asm__ __volatile__("mcr p15, 0, %0, c1, c0, 0" : : "r"(reg_val));
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:615]: __set_SCTLR

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 5]
```diff
@@ -0,0 +1,800 @@
...
+		*phys = (uintptr_t)pa_resolved;
+	}
+	return rc;
+}
+
+static void invalidate_tlb_all(void)
+{
+	uint32_t reg_val = 0; /* 0 = opc2 = invalidate entire TLB */
+	__asm__ __volatile__("mcr p15, 0, %0, c8, c7, 0" : : "r"(reg_val));
+	__asm__ __volatile__("dsb sy");
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:798]: __DSB()

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 6]
```diff
@@ -0,0 +1,800 @@
...
+	__asm__ __volatile__("isb");
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:799]: __ISB()

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 7]
```diff
@@ -0,0 +1,800 @@
...
+	if (phys) {
+		*phys = (uintptr_t)pa_resolved;
+	}
+	return rc;
+}
+
+static void invalidate_tlb_all(void)
+{
+	uint32_t reg_val = 0; /* 0 = opc2 = invalidate entire TLB */
+	__asm__ __volatile__("mcr p15, 0, %0, c8, c7, 0" : : "r"(reg_val));
```
Comment (bbolen) [inline arch/arm/core/aarch32/mmu/arm_mmu.c:797]: __set_TLBIALL(0)

**Review by bbolen — CHANGES_REQUESTED** [review]

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 1 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c:601]: CMSIS integration is my goal, yet, with CMSIS being an external component, I haven't yet figured out where to commit the CPU-specific header core_ca9.h, similar to the already existing core_cr5.h, as CMSIS header inclusion entry point.

Comment (stephanosio) [reply arch/arm/core/aarch32/mmu/arm_mmu.c:601]: CMSIS-Core(A) has already been integrated.

See https://github.com/zephyrproject-rtos/cmsis/tree/master/CMSIS/Core_A/Include

File: arch/arm/core/aarch32/mmu/arm_mmu.c [hunk 8]
```diff
@@ -0,0 +1,800 @@
...
+		 "entries value is %u\n",
+		 va, arm_mmu_l2_next_free_table,
+		 l2_page_tables_status[arm_mmu_l2_next_free_table].entries);
+
+	l2_page_tables_status[arm_mmu_l2_next_free_table].l1_index =
+		((va >> ARM_MMU_PTE_L1_INDEX_PA_SHIFT) & ARM_MMU_PTE_L1_INDEX_MASK);
+	l2_page_tables_status[arm_mmu_l2_next_free_table].entries = 0;
+	l2_page_table = &l2_page_tables[arm_mmu_l2_next_free_table];
+
+	if (--arm_mmu_l2_tables_free > 0)
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c:108]: Correct, that wasn't a clever shortcut...

File: soc/arm/xilinx_zynq7000/soc.c [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @brief Basic hardware initialization of the Zynq-7000 SoC
+ *
+ * Performs the basic initialization of the Zynq-7000 SoC.
+ *
+ * @return 0
+ */
+static int soc_xlnx_zynq7000_init(const struct device *arg)
+{
+	uint32_t reg_val;
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply soc/arm/xilinx_zynq7000/soc.c:76]: I'm leaving this in for now - I already had an implementation in here that writes the registers which configure where the 256 kB of OCM are located, based on the respective 'chosen' entry in the device tree. The power-on configuration is 3/4 of the OCM starting at address 0, 1/4 at 0xFFFC0000. The attempt to move the 64k located at the high address to the low address currently crashes the system, as we neither have proper SMP yet nor have I looked into how to shut down the 2nd core in non-SMP mode yet. The 'wait for event' loop code the 2nd core idles in after power on is located in that particular 64k block that's located at the high address by default. Therefore, I removed this code for now, but it shall eventually return one day.

File: drivers/serial/uart_xlnx_ps.c [hunk 10]
```diff
@@ -845,20 +842,19 @@ static int uart_xlnx_ps_fifo_fill(const struct device *dev,
...
-	reg_base = dev_cfg->uconf.regs;
-	reg_val = sys_read32(reg_base + XUARTPS_SR_OFFSET);
-	while (onum < size && (reg_val & XUARTPS_SR_TXFULL) == 0) {
-		sys_write32((uint32_t)(tx_data[onum] & 0xFF),
-				reg_base + XUARTPS_FIFO_OFFSET);
-		onum++;
-		reg_val = sys_read32(reg_base + XUARTPS_SR_OFFSET);
+
+	sys_write32(XUARTPS_IXR_TXEMPTY, reg_base + XUARTPS_IDR_OFFSET);
+	while (size) {
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/serial/uart_xlnx_ps.c]: done

File: drivers/serial/uart_xlnx_ps.c [hunk 11]
```diff
@@ -845,20 +842,19 @@ static int uart_xlnx_ps_fifo_fill(const struct device *dev,
...
+		while ((sys_read32(reg_base + XUARTPS_SR_OFFSET) & XUARTPS_SR_TXFULL) != 0) {}
+		sys_write32((uint32_t)(tx_data[data_iter++] & 0xFF), reg_base + XUARTPS_FIFO_OFFSET);
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/serial/uart_xlnx_ps.c]: done

File: soc/arm/xilinx_zynq7000/soc.c [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+			      0xF8000000,
+			      0x1000,
+			      MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),
+	MMU_REGION_FLAT_ENTRY("MPCore",
+			      0xF8F00000,
+			      0x2000,
+			      MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),
+	/* ARM Arch timer, GIC are covered by the MPCore mapping */
+/* UARTs */
+#if DT_NODE_HAS_STATUS(DT_INST(0, xlnx_xuartps), okay)
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply soc/arm/xilinx_zynq7000/soc.c]: done

File: boards/arm/qemu_cortex_a9/Kconfig.defconfig [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	int
+	default 111111111
+
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
+	
+if SERIAL
+
+	config SERIAL_HAS_DRIVER
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply boards/arm/qemu_cortex_a9/Kconfig.defconfig]: removed that item

File: boards/arm/qemu_cortex_a9/Kconfig.defconfig [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	default y
+
+	config UART_INTERRUPT_DRIVEN
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply boards/arm/qemu_cortex_a9/Kconfig.defconfig]: That's where that item is now. Yet I'm currently wondering how I came up with the combination of CONSOLE=y, INTERRUPT_DRIVEN=n... More testing of the QEMU target is coming up anyways.

File: boards/arm/qemu_cortex_a9/qemu_cortex_a9.dts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+	sram0: memory@100000 {
+		compatible = "mmio-sram";
+		reg = <0x00100000 DT_SIZE_M(512)>;
+	};
+
+	flash0: flash@0 {
+		reg = <0x00000000 DT_SIZE_M(256)>;
+	};
+
+	aliases {
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply boards/arm/qemu_cortex_a9/qemu_cortex_a9.dts]: ... aaaaand it's gone!

File: boards/arm/qemu_cortex_a9/qemu_cortex_a9_defconfig [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+CONFIG_QEMU_ICOUNT=y
+CONFIG_QEMU_ICOUNT_SHIFT=3
+
+CONFIG_ARM_ARCH_TIMER=y
+
+CONFIG_SERIAL=y
+CONFIG_CONSOLE=y
+CONFIG_NETWORKING=y
+CONFIG_NET_QEMU_ETHERNET=y
+CONFIG_GPIO=n
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply boards/arm/qemu_cortex_a9/qemu_cortex_a9_defconfig]: done

File: include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld [hunk 17]
```diff
@@ -84,10 +85,11 @@ _region_min_align = 4;
...
 MEMORY
 {
-    FLASH    (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
-    SRAM     (wx) : ORIGIN = RAM_ADDR, LENGTH = RAM_SIZE
+    FLASH     (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
+    SRAM      (wx) : ORIGIN = RAM_ADDR, LENGTH = RAM_SIZE
+    DT_REGION_FROM_NODE_STATUS_OKAY(OCM, rw, DT_CHOSEN(zephyr_ocm))
     /* Used by and documented in include/linker/intlist.ld */
-    IDT_LIST (wx) : ORIGIN = (RAM_ADDR + RAM_SIZE), LENGTH = 2K
+    IDT_LIST  (wx) : ORIGIN = (RAM_ADDR + RAM_SIZE), LENGTH = 2K
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld]: not sure what happened there, but it's fixed...

File: include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld [hunk 18]
```diff
@@ -84,10 +85,11 @@ _region_min_align = 4;
 
 MEMORY
 {
-    FLASH    (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
-    SRAM     (wx) : ORIGIN = RAM_ADDR, LENGTH = RAM_SIZE
+    FLASH     (rx) : ORIGIN = ROM_ADDR, LENGTH = ROM_SIZE
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply include/arch/arm/aarch32/cortex_a_r/scripts/linker.ld]: dito

File: drivers/timer/Kconfig [hunk 19]
```diff
@@ -121,6 +121,18 @@ DT_COMPAT_ARM_V7M_SYSTICK := arm,armv7m-systick
 DT_COMPAT_ARM_V8M_SYSTICK := arm,armv8m-systick
 DT_COMPAT_ARM_V8_1M_SYSTICK := arm,armv8.1m-systick
 
+config ARM_ARCH_TIMER_ERRATUM_740657
+	bool "ARM architected timer is affected by ARM erratum 740657"
+	default n
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/timer/Kconfig]: done

File: include/linker/sections.h [hunk 20]
```diff
@@ -51,6 +51,10 @@
 #define _DTCM_DATA_SECTION_NAME	.dtcm_data
 #define _DTCM_BSS_SECTION_NAME		.dtcm_bss
 #define _DTCM_NOINIT_SECTION_NAME	.dtcm_noinit
+
+#define _OCM_DATA_SECTION_NAME	.ocm_data
+#define _OCM_BSS_SECTION_NAME		.ocm_bss
+#define _OCM_NOINIT_SECTION_NAME	.ocm_noinit
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply include/linker/sections.h:57]: The Ethernet driver places its DMA area in the bss section, noinit is not really required. The availability of those three sections is simply based on what other On-Chip memory types that are already supported offer.

File: drivers/timer/arm_arch_timer.c [hunk 21]
```diff
@@ -39,7 +57,28 @@ static void arm_arch_timer_compare_isr(const void *arg)
...
 		arm_arch_timer_set_irq_mask(true);
+#ifdef CONFIG_ARM_ARCH_TIMER_ERRATUM_740657
+		/*
+		 * In tickless mode, the compare register is normally not
+		 * updated from within the ISR. Yet, to work around the timer's
+		 * erratum, a new value *must* be written while the interrupt
+		 * is being processed before the interrupt is acknowledged
+		 * by the handling interrupt controller.
+		 */
+		arm_arch_timer_set_compare(0xFFFFFFFFFFFFFFFFLLU);
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply drivers/timer/arm_arch_timer.c]: done

File: include/arch/arm/aarch32/cortex_a_r/timer.h [hunk 22]
```diff
@@ -63,6 +63,30 @@ static ALWAYS_INLINE void arm_arch_timer_set_compare(uint64_t val)
 	sys_write32(ctrl, TIMER_REG_GET(TIMER_CTRL));
 }
 
+#if defined(CONFIG_ARM_ARCH_TIMER_ERRATUM_740657) || defined(CONFIG_QEMU_TARGET)
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply include/arch/arm/aarch32/cortex_a_r/timer.h:68]: I did actually observe a malfunctioning timer interrupt in QEMU if no new compare value was written within the ISR in tickless mode.

File: include/arch/arm/aarch32/cortex_a_r/timer.h [hunk 23]
```diff
@@ -63,6 +63,30 @@ static ALWAYS_INLINE void arm_arch_timer_set_compare(uint64_t val)
...
+ * This functionality is not present in the aarch64 implementation of the
+ * ARM global timer access functions. 
+ * 
+ * comp. ARM Cortex-A9 processors Software Developers Errata Notice,
+ * ARM document ID032315.
+ */
+
+static ALWAYS_INLINE uint8_t arm_arch_timer_get_int_status(void)
+{
+	return (uint8_t)(sys_read32(TIMER_REG_GET(TIMER_ISR)) & 0x1);
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply include/arch/arm/aarch32/cortex_a_r/timer.h]: done, it's now called TIMER_ISR_EVENT_FLAG.

File: arch/arm/core/aarch32/prep_c.c [hunk 24]
```diff
@@ -182,7 +183,8 @@ void z_arm_prep_c(void)
 #endif
 	z_bss_zero();
 	z_data_copy();
-#if defined(CONFIG_ARMV7_R) && defined(CONFIG_INIT_STACKS)
+#if (defined(CONFIG_ARMV7_R) && defined(CONFIG_INIT_STACKS)) \
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply arch/arm/core/aarch32/prep_c.c]: done

File: arch/arm/core/aarch32/mmu/arm_mmu_priv.h [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+};
+
+/*
+ * Data structure used to describe memory areas defined by the
+ * current Zephyr image, for which an identity mapping (pa = va)
+ * will be set up. Those memory areas are processed during the
+ * MMU initialization.
+ */
+struct arm_mmu_flat_range {
+	char *name;
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu_priv.h]: done

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 1 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c:603]: When looking for the CMSIS, I only looked within zephyrproject-rtos/zephyr and in ARM-software/CMSIS_5, but not zephyrproject-rtos/... D'oh!

**Comment by evgeniy-paltsev** [comment]
@ibirnbaum may I ask you to rebase your changes on latest Zephyr version?
As I can see now from 'Files changed' tab you are deleting/changing a lot of ARC code which looks really strange :)

**Comment by ibirnbaum** [comment]
@evgeniy-paltsev Something must have gone wrong yesterday. I had some conflicts due to certain files having been updated in the meantime, I merged, added & pushed the updated versions, the web frontend was happy with that, but the CI kept complaining that there was a merge conflict it couldn't resolve. I'm just trying to get rid of the merge commits that caused this issue. Not too sure about the chances of success, though...

**Comment by ibirnbaum** [comment]
@jackrosenthal Thanks for cleaning up the exploding list of reviewers after that one broken merge that eventually got pushed onto my branch! To all those who didn't come for aarch32 stuff: move along, there's nothing to see here ;-)

The use of CMSIS in the MMU code is in progress, and initial runs of the kernel testsuite on the QEMU target didn't look too bad, but I did observe the occasional strange behaviour which I attribute to QEMU rather than the code, as it runs fine on actual hardware. It has been stated that the Xilinx targets integrated in vanilla QEMU aren't that great in terms of quality & stability, and that the way to go for such targets is the Xilinx derivative of QEMU in conjunction with a device tree rather than a pre-defined target. This is exactly how the Cortex-R5 QEMU board is set up, and therefore, the Xilinx derivative of QEMU is already a part of the SDK. Although branded arm64, this version also supports the Zynq, so I'm currently working on generating a device tree in order to have a reliable QEMU represenatation of the Cortex-A9.

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 3 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: done

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 4 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: done

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 5 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: done

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 6 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: done

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 7 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: done

File: boards/arm/qemu_cortex_a9/Kconfig.defconfig (hunk 14 above)
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply boards/arm/qemu_cortex_a9/Kconfig.defconfig]: The config files of the QEMU A9 target have been re-worked.

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 2 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: done

File: arch/arm/core/aarch32/mmu/arm_mmu.c (hunk 1 above)
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu.c]: the MMU code has been modified to use CMSIS whereever supported, the only exceptions being TTBR1 and TTBCR, which aren't covered by cmsis_cp15.h.

File: include/linker/sections.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply include/linker/sections.h]: Removed the noinit-related define, as this actually wasn't considered in the linker command file.

File: arch/arm/core/aarch32/mmu/arm_mmu_priv.h [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+	uint32_t ns		: 1;
+	uint32_t s		: 1;
+	uint32_t tex		: 3;
+	uint32_t xn		: 1;
+	uint32_t id_mask	: 2;
+	uint32_t domain		: 4;
+	uint32_t reserved	: 15;
+};
+
+static void arm_mmu_release_l2_table(struct arm_mmu_l2_page_table* l2_page_table);
```
[the remark this answers is not in the document]
Comment (ibirnbaum) [reply arch/arm/core/aarch32/mmu/arm_mmu_priv.h]: done.

**Review by bbolen — CHANGES_REQUESTED** [review]
Most of the stuff is nitpicking and I'm not necessarily asking for changes on those.  The mmu free table handling is the thing that I think needs to be looked at.

**Review by bbolen — CHANGES_REQUESTED** [review]

**Review by bbolen — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
This looks good to me. The only question I had revolved around having some sort of common `CORTEX_R_A32` Kconfig option to avoid the duplication of the `||`ing but honestly I cannot come up with a good name or description for it.

**Comment by cfriedt** [comment]
@ibirnbaum, @carlescufi  - this might have introduced a new bug and regression into main. Is there a fix on the way?
```
twister -i -p qemu_cortex_a53_smp -s tests/kernel/fatal/exception/kernel.common.stack_sentinel
twister -i -p qemu_cortex_a53_smp  -s tests/kernel/smp/kernel.multiprocessing.smp
twister -i -p qemu_cortex_a53_smp  -s tests/kernel/sched/schedule_api/kernel.scheduler
twister -i -p qemu_cortex_a9 -s tests/kernel/fatal/exception/kernel.common.stack_sentinel
```

**Pull request merged**